File: Heap/502. IPO.py
# 2024/06/15
# Solution Top 1 的解法，牛逼！
class Solution(object):
    def findMaximizedCapital(self, k, w, profits, capital):
        """
        :type k: int
        :type w: int
        :type profits: List[int]
        :type capital: List[int]
        :rtype: int
        """
        # 可以一口氣放兩個組，使用 capital 來排序
        projects = [(capital[i], profits[i]) for i in range(len(profits))]
        projects.sort()

        maxHeap = []
        i = 0
        for _ in range(k):
            while(i < len(projects) and projects[i][0] <= w):
                    # 因為 heapq 由小到大排序，所以 push -projects[i][1]，這樣最大的反而最小，等等 w 再用 - 的讓他變回正數
                    heapq.heappush(maxHeap, -projects[i][1])
                    i += 1
            # if not <list name> 可以確認 list 是否 empty
            if not maxHeap:
                break
            
            w -= heapq.heappop(maxHeap)

        return w

File: Heap/502. IPO.cpp
// 2024/06/22
class Solution {
    public:
        
        struct Project{
            int profit;
            int capital;

            // 由小到大
            bool const operator < (Project &item) const{
                return capital < item.capital;
            }
        };
        int findMaximizedCapital(int k, int w, vector<int>& profits, vector<int>& capital) {
            int n = profits.size();
            Project projects[n];

            for(int i = 0; i < n; ++i){
                projects[i].profit = profits[i];
                projects[i].capital = capital[i];
            }
            sort(projects, projects + n);

            int cnt = 0;
            int ansCapital = w;
            priority_queue<int> pq;
            for(int i = 0; i < k; ++i){
                while(cnt < n && projects[cnt].capital <= ansCapital){
                    pq.push(projects[cnt].profit);
                    cnt++;
                }

                if(pq.empty()) break;

                ansCapital += pq.top();
                pq.pop();
            }

            return ansCapital;
        }
};